### PR TITLE
Implement Usort hints

### PR DESCRIPTION
**Pull request by pefontana**
Implement hints of the Usort module
* USORT_ENTER_SCOPE
* USORT_BODY
* USORT_VERIFY
* USORT_VERIFY_MULTIPLICITY_ASSERT
* USORT_VERIFY_MULTIPLICITY_BODY

Implement Sort of  []lambdaworks.Felt

### CODE REVIEW

File: cairo_programs/proof_programs/usort.cairo [hunk 1]
```diff
@@ -0,0 +1,22 @@
+%builtins range_check
```
Comment (jrchatruc) [inline cairo_programs/proof_programs/usort.cairo]: Move this file to be a symlink to the one under the `cairo_programs` directory

File: pkg/hints/usort_hints.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+func (s SortFelt) Less(i, j int) bool {
+	a, b := s[i], s[j]
+
+	return a.Cmp(b) == -1
+}
+
+// Implements hint:
+// %{ vm_enter_scope(dict(__usort_max_size = globals().get('__usort_max_size'))) %}
+func usort_enter_scope(executionScopes *types.ExecutionScopes) error {
+	usort_max_size_interface, err := executionScopes.Get("usort_max_size")
```
Comment (jrchatruc) [inline pkg/hints/usort_hints.go:30]: Use camelCase for variable and function names (there are a couple other instances of this throughout the code)

File: pkg/hints/usort_hints.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+	for key := range output {
+		multiplicities = append(multiplicities, uint64(len(positions_dict[output[key]])))
+	}
+
+	for i := range multiplicities {
+		vm.Segments.Memory.Insert(multiplicities_base.AddUint(uint(i)), memory.NewMaybeRelocatableFelt(lambdaworks.FeltFromUint64(multiplicities[i])))
+	}
+	ids.Insert("output", memory.NewMaybeRelocatableRelocatable(output_base), vm)
+	ids.Insert("multiplicities", memory.NewMaybeRelocatableRelocatable(multiplicities_base), vm)
```
Comment (fmoletta) [inline pkg/hints/usort_hints.go]: Please handle the errors returned by ids.Insert

File: pkg/hints/usort_hints.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+	multiplicities_base := vm.Segments.AddSegment()
+
+	multiplicities := make([]uint64, 0, len(output))
+
+	for key := range output {
+		multiplicities = append(multiplicities, uint64(len(positions_dict[output[key]])))
+	}
+
+	for i := range multiplicities {
+		vm.Segments.Memory.Insert(multiplicities_base.AddUint(uint(i)), memory.NewMaybeRelocatableFelt(lambdaworks.FeltFromUint64(multiplicities[i])))
```
Comment (fmoletta) [inline pkg/hints/usort_hints.go]: Please handle the error returned by Memory.Insert

File: pkg/hints/usort_hints.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+	output := make([]lambdaworks.Felt, 0, len(positions_dict))
+
+	for key := range positions_dict {
+		output = append(output, key)
+	}
+
+	sort.Sort(SortFelt(output))
+
+	output_len := memory.NewMaybeRelocatableFelt(lambdaworks.FeltFromUint64(uint64((len(output)))))
+	ids.Insert("output_len", output_len, vm)
```
Comment (fmoletta) [inline pkg/hints/usort_hints.go]: Please handle the error returned by ids.Insert

File: pkg/hints/usort_hints.go [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+	output_base := vm.Segments.AddSegment()
+
+	for i := range output {
+		vm.Segments.Memory.Insert(output_base.AddUint(uint(i)), memory.NewMaybeRelocatableFelt(output[i]))
```
Comment (fmoletta) [inline pkg/hints/usort_hints.go]: Please handle the error returned by Memory.Insert

File: pkg/hints/usort_hints.go [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+	}
+
+	positions, cast_ok := positions_interface.([]uint64)
+
+	if !cast_ok {
+		return errors.New("Error casting positions to []uint64")
+	}
+
+	if len(positions) != 0 {
+		return errors.New("unexpected verify multiplicity fail: positions length != 0")
```
Comment (fmoletta) [inline pkg/hints/usort_hints.go]:
```suggestion
		return errors.New("Assertion failed: len(positions) == 0 ")
```

File: pkg/hints/usort_hints.go [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+	if !cast_ok {
+		return errors.New("Error casting last_pos to uint64")
+	}
+
+	current_pos := positions[len(positions)-1]
+
+	executionScopes.AssignOrUpdateVariable("positions", positions[:len(positions)-1])
+
+	next_item_index := current_pos - last_pos
+	ids.Insert("next_item_index", memory.NewMaybeRelocatableFelt(lambdaworks.FeltFromUint64(next_item_index)), vm)
```
Comment (fmoletta) [inline pkg/hints/usort_hints.go]: Please handle the error returned by ids.Insert

File: pkg/hints/usort_hints_test.go [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+	. "github.com/lambdaclass/cairo-vm.go/pkg/hints"
+	. "github.com/lambdaclass/cairo-vm.go/pkg/hints/hint_codes"
+	. "github.com/lambdaclass/cairo-vm.go/pkg/hints/hint_utils"
+	"github.com/lambdaclass/cairo-vm.go/pkg/lambdaworks"
+	"github.com/lambdaclass/cairo-vm.go/pkg/types"
+	. "github.com/lambdaclass/cairo-vm.go/pkg/vm"
+	. "github.com/lambdaclass/cairo-vm.go/pkg/vm/memory"
+)
+
+func TestReverseFeltArray(t *testing.T) {
```
Comment (fmoletta) [inline pkg/hints/usort_hints_test.go]:
```suggestion
func TestSortFeltArray(t *testing.T) {
```

File: cairo_programs/proof_programs/usort.cairo (hunk 1 above)
Comment (pefontana) [reply cairo_programs/proof_programs/usort.cairo]: Done [fe15372](https://github.com/lambdaclass/cairo-vm_in_go/pull/244/commits/fe1537281335b8887bd6dd42d6cb1824f7b63c41)

File: pkg/hints/usort_hints.go (hunk 2 above)
Comment (pefontana) [reply pkg/hints/usort_hints.go:30]: hahaha 
Done [2d0f3e8](https://github.com/lambdaclass/cairo-vm_in_go/pull/244/commits/2d0f3e8134a59a8cedbc49c98bcdb48d9c6e4418)

File: pkg/hints/usort_hints.go (hunk 5 above)
Comment (pefontana) [reply pkg/hints/usort_hints.go]: Nice catch @fmoletta !
Done [6849011](https://github.com/lambdaclass/cairo-vm_in_go/pull/244/commits/6849011c47ff79b42367713e14a6e283f62b8310)

File: pkg/hints/usort_hints.go (hunk 6 above)
Comment (pefontana) [reply pkg/hints/usort_hints.go]: Done
[fc6134f](https://github.com/lambdaclass/cairo-vm_in_go/pull/244/commits/fc6134f314b97ed3435de42c78e75cf28919082a)

**Review by fmoletta — APPROVED** [review]

**Review by jrchatruc — APPROVED** [review]

**Pull request merged**